[Threading] Fix threading package defaults (take 2)

al45tair opened the pull request:
These are better done via the SwiftConfigureSDK mechanism rather than how I was doing them previously.  Additionally, I've changed the way that the swift-threading-package option works.  In addition to specifying just a single package name, you can specify it as a CMake list (i.e. separated by semicolons) of colon-separated `sdk:package` pairs, e.g. `osx:darwin;linux:pthreads`.  You can also override it for all SDKs and then specify for a given SDK; specifications for a particular SDK take precedence over the global override.  For instance `pthreads;osx:darwin` says to use `pthreads` except on the OS X SDK where we should use `darwin`.

al45tair committed:
[Threading] Fix threading package defaults.

These are better done via the SwiftConfigureSDK mechanism rather than
how I was doing them previously.  Additionally, I've changed the way
that the swift-threading-package option works.  In addition to
specifying just a single package name, you can specify it as a CMake
list (i.e.  separate by semicolons) of colon-separated `sdk:package`
pairs, e.g.  `osx:darwin;linux:pthreads`.  You can also override it
for all SDKs and then specify for a given SDK; specifications for a
particular SDK take precedence over the global override.  For instance
`pthreads;osx:darwin` says to use `pthreads` except on the OS X SDK
where we should use `darwin`.

al45tair commented:
@swift-ci Please test

al45tair commented:
preset=stdlib_S_standalone_minimal_macho_x86_64,build,test
@swift-ci please test with toolchain and preset

al45tair commented:
@swift-ci Please test

al45tair commented:
preset=stdlib_S_standalone_minimal_macho_x86_64,build,test
@swift-ci please test with toolchain and preset

buttaface commented on utils/build-script-impl:
```diff
@@ -2227,8 +2227,8 @@ for host in "${ALL_HOSTS[@]}"; do
 
-                if [[ "${SWIFT_THREADING_PACKAGE}" ]] ; then
-                    case "${SWIFT_THREADING_PACKAGE}" in
+                if [[ "${SWIFT_STDLIB_THREADING_PACKAGE}" ]] ; then
```
If this is an empty string, I believe bash evaluates this as false and skips this block, so you'll need to correct the list of  cases below.

al45tair replied:
I don't believe this is relevant any more either.

finagolfin reviewed (changes requested):
Overall, looks good, but a few nits.

al45tair commented:
> Overall, looks good, but a few nits.

:-) I wasn't quite ready for it to be reviewed yet, but thanks for the comments — they were a helpful reminder that I needed to revisit a few things first.

al45tair commented on cmake/modules/AddSwiftUnittests.cmake:
```diff
@@ -54,7 +53,11 @@ function(add_swift_unittest test_dirname)
   endif()
-  threading_package_name("${SWIFT_HOST_VARIANT_SDK}" _threading_package)
+  if(SWIFT_STDLIB_THREADING_PACKAGE)
+    string(TOUPPER "${SWIFT_STDLIB_THREADING_PACKAGE}" _threading_package)
```
There's a complication with the new mechanism in that setting it for an SDK affects both toolchain and stdlib compilations if they happen to be using the same SDK. This was the first thing that came to mind to work around that, but of course it means the name is wrong. But I have something much better now.

al45tair commented on stdlib/cmake/modules/AddSwiftStdlib.cmake:
```diff
@@ -355,7 +354,11 @@ function(_add_target_variant_c_compile_flags)
 
-  threading_package_name("${CFLAGS_SDK}" _threading_package)
+  if(SWIFT_STDLIB_THREADING_PACKAGE)
+    string(TOUPPER "${SWIFT_STDLIB_THREADING_PACKAGE}" _threading_package)
```
These are gone anyway. I've changed the way the override works.

al45tair commented:
@swift-ci Please test

buttaface commented:
Nice, thanks for adding the list of overrides, that's useful for cross-compilation.

buttaface commented on utils/build-script-impl:
```diff
@@ -206,7 +206,7 @@ KNOWN_SETTINGS=(
     swift-stdlib-supports-backtrace-reporting     ""                "whether to build stdlib assuming the runtime environment provides the backtrace(3) API, if not set defaults to true on all platforms except for Cygwin, Haiku and wasm"
     swift-runtime-static-image-inspection         "0"               "whether to build stdlib assuming the runtime environment only supports a single runtime image with Swift code"
-    swift-threading-package                       ""                "which threading package to use for Swift; valid values are empty string (default based on platform), 'pthreads', 'darwin', 'linux', 'win32', 'c11', 'none'"
+    swift-threading-package                  ""                     "override the threading package for the host build; this is either a single package or a semicolon-separated list of sdk:package pairs.  Valid packages are empty string (no override), 'pthreads', 'darwin', 'linux', 'win32', 'c11', 'none'"
```
Nit, move the quotes so it lines up with the other values in one column.

buttaface commented on utils/build-script-impl:
```diff
@@ -206,7 +206,7 @@ KNOWN_SETTINGS=(
     swift-stdlib-supports-backtrace-reporting     ""                "whether to build stdlib assuming the runtime environment provides the backtrace(3) API, if not set defaults to true on all platforms except for Cygwin, Haiku and wasm"
     swift-runtime-static-image-inspection         "0"               "whether to build stdlib assuming the runtime environment only supports a single runtime image with Swift code"
-    swift-threading-package                       ""                "which threading package to use for Swift; valid values are empty string (default based on platform), 'pthreads', 'darwin', 'linux', 'win32', 'c11', 'none'"
+    swift-threading-package                       ""                     "override the threading package for the host build; this is either a single package or a semicolon-separated list of sdk:package pairs.  Valid packages are empty string (no override), 'pthreads', 'darwin', 'linux', 'win32', 'c11', 'none'"
```
Heh, now the last column with the description is misaligned. 😉

al45tair replied:
😂 Fixed now.

al45tair commented:
@swift-ci Please test

buttaface commented:
@compnerd, would you review?

compnerd commented on cmake/modules/SwiftConfigureSDK.cmake:
```diff
@@ -101,6 +102,34 @@ function(remove_sdk_unsupported_archs name os sdk_path architectures_var)
+    if(elt_list_len EQUAL 1)
+      list(GET elt_list 0 global_override)
+      string(TOLOWER "${global_override}" global_override)
+    else()
```
Should we have a case to ensure that someone doesn't form an invalid input with multiple `:`?

al45tair replied:
Good idea. Done.

compnerd commented on cmake/modules/SwiftConfigureSDK.cmake:
```diff
@@ -101,6 +102,34 @@ function(remove_sdk_unsupported_archs name os sdk_path architectures_var)
 endfunction()
 
+# Work out which threading package to use by consulting SWIFT_THREADING_PACKAGE
+function(find_threading_package sdk default package_var)
```
`get_threading_package` seems like a better name as it is meant to retrieve the value.

al45tair committed:
[Threading] Minor improvements to SWIFT_THREADING_PACKAGE support

Added a check that someone hasn't passed us an argument with too many
colons.  Also renamed `find_threading_package` to `get_threading_package`
to better reflect what it does.

al45tair commented:
@swift-ci Please test

al45tair commented:
@swift-ci Please test macOS platform

compnerd reviewed (approved):
Seems reasonable to me.  I didn't carefully review the bash bits, and I suppose it is fine as is, though I tend to prefer to write `sdk=${elt%%:*}` and `package=${elt##*:}` for longest prefix and suffix replacement rather than the regular substitution.
